### DISCUSSION TITLE
Multiple domains with different mail server FQDN

### DISCUSSION THREAD

**Discussion by dweipert-3138720606** [docker-mailserver/docker-mailserver#3700] [Q&A] [answered]
I know that docker-mailserver supports multiple domains, but those domains would have to point to the same FQDN MX record then.
Could I theoretically run multiple container instances with different Mailserver-FQDN from the same volumes?
So configuration files are the same, but the `hostname` and `environment.SSL_DOMAIN` values would be different.

Only problem I see right now would be the published ports with multiple containers, but with traefik I should be able to route them with a normal `Host` rule.

Thoughts?

**Comment by polarathene** [comment, marked as answer]
> Could I theoretically run multiple container instances with different Mailserver-FQDN from the same volumes?

Possibly? It might depend on the configuration.

I am not sure it would be a good idea in some cases for some data if there's an assumption of it being managed by one owner/process. An unlikely example would be using our `setup ...` commands which modify config files, those could potentially be automated via some custom API layer for example and triggered at roughly the same time that one overwrites the other or one is writing to a file while another is reading it which could result in reading incomplete data 🤷‍♂️ 

More realistically, rather than our own configs, it's more of a concern with the services running in the container. We have state volumes which absolutely shouldn't be shared across containers, and I'd say the same for the log volume. If there's no overlap with the mail storage volume that may be ok, just keep in mind that the VMAIL user (Dovecot, currently `docker:docker` / `5000:5000`) would need to be in sync, as if you change it for one container it'll fight with the other back and forth over changing ownership for the data.

For our config volume, you have our change detection service that would trigger in both containers when they monitor a change AFAIK. I am aware of some users that have run multiple instances of DMS with this storage shared, so it may work?

---

> I know that docker-mailserver supports multiple domains, but those domains would have to point to the same FQDN MX record then.

Just to clarify, this does not have to any relation to the supported mail domains.

- DMS runs at `mail.example.com`
- Client 1 has the mail accounts setup for `@client-a.com`
- Client 2 has their mail accounts setup for `@client-b.com`
- Both clients have their DNS set to point to DMS at `mail.example.com`, and there is no `@example.com` mail account managed, just a mail server.

The next area that needs to be clarified (_and we've really tried to communicate these concerns better with the v13 docs_), the SSL / TLS certificates are not for mail domains, but the DMS FQDN.

You only need the certificate for `mail.example.com`, aka DMS. None needed for the mail domains it manages.

Given the above, is there a use-case you're wanting to support where you're trying to keep the same volumes/config, but running different instances for a different DMS FQDN to support?

- rDNS would not map to the different container assigned FQDN unless you have multiple IP addresses to the server, which if not a concern for you, the client DNS records I believe can just point their own MX records to the server IP running DMS. MX should just be for receiving.
- I recall Traefik support for SMTP traffic being a little iffy, that might have been related to establishing TLS instead of using TLS passthrough to DMS.
- The hostname configured for the DMS container then is used for Postfix to recognize it's FQDN to operate with. This is relevant for:
  - The certificate and typically internal mail (_notifications like updates of new DMS releases to the `root` user, anything related to internal services like Amavis/SA, Fail2Ban, etc, Postmaster if anything went wrong with Postfix, eg used for bounce mail sent back if delivery fails IIRC_).
  - Some users choose to deploy with overlapping the hostname/FQDN with the mail domain (_our docs FAQ refers to this as a bare domain, but we don't encourage it, eg both use `example.com`, no `mail` subdomain for DMS_). 
  - It is what is used as the HELO/EHLO hostname when sending mail to other MTAs, this would be looked up to verify SPF and other DNS records, that DMS can be trusted to send mail on behalf of the sender address mail domain used. Unrelated to MX record, but often the same.

---

> Only problem I see right now would be the published ports with multiple containers, but with traefik I should be able to route them with a normal `Host` rule.

Just be sure to preserve the source address correctly. If Postfix and other services only see the Traefik IP as the source, it can be problematic to correct operation. From what I have seen from other users deploying with Traefik there is some additional configuration for DMS to support that, but you're perhaps already aware of that :)

**Reply by dweipert-3138720606** [reply]
Thanks for your thoughts!

I wanted to reduce the potential risk of spammers targeting my mail server domain directly.
That's why I thought separating the mail server for both of my domains would be a good idea.
Although it would still be the same server IP, so I don't know whether I would win anything with that.

For now I abandoned that idea and just went with one instance for both domains.

Regarding the Traefik IP, I wasn't aware of that. My current setup just uses traefik for the SSL certificate. The docker-mailserver container itself publishes the ports directly to the host.
If I ever pick up the idea of multiple instances again, I will look into that, thanks!
